Add Turks and Caicos Islands holidays

ShalomDee opened the pull request:
<!--
  Thanks for contributing to holidays!
-->

## Proposed change

This PR adds official public holiday support for the **Turks and Caicos Islands (TC)** to the `holidays` package.

**Issue:** Fixes #2435

Holidays included are based on:
- https://en.wikipedia.org/wiki/Public_holidays_in_the_Turks_and_Caicos_Islands
- https://destinationtci.tc/turks-and-caicos-islands-public-holidays/
- https://www.timeanddate.com/holidays/turks-and-caicos-islands/

The implementation covers fixed holidays and movable holidays like Easter-related dates and the King's Birthday.

## Type of change

- [x] New country/market holidays support (thank you!)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've successfully run `make check`, all checks and tests are green

<!--
  Thanks again for your contribution!
-->

[contributing-guidelines]: https://github.com/vacanza/holidays/blob/dev/CONTRIBUTING.md
[docs]: https://github.com/vacanza/holidays/tree/dev/docs/source

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,81 @@
+        # JAGS McCartney Day (last Monday in May)
+        self._add_holiday_last_mon_of_may(tr("JAGS McCartney Day"))
+
+        # King's Birthday (second Monday in June)
+        self._add_holiday_2nd_mon_of_jun(tr("King's Birthday"))
```
Until 2023, this holiday was named "Queen's Birthday".

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,81 @@
+        # National Youth Day (last Friday in September)
+        self._add_holiday_last_fri_of_sep(tr("National Youth Day"))
+
+        # National Heritage Day (second Monday in October)
+        self._add_holiday_2nd_mon_of_oct(tr("National Heritage Day"))
```
Earlier this holiday was named Columbus Day (we need to find out when the name was changed).

KJhellico commented on holidays/locale/en_US/LC_MESSAGES/TC.po:
```diff
@@ -0,0 +1,76 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Holidays 0.22\n"
+"POT-Creation-Date: 2023-03-23 14:44+0200\n"
+"PO-Revision-Date: 2023-03-23 14:47+0200\n"
+"Last-Translator: Your Name <your.email@example.com>\n"
```
Here should be version 0.71 (currently), the latest dates, and your name and email.

KJhellico commented on holidays/registry.py:
```diff
@@ -174,6 +174,7 @@
     "tonga": ("Tonga", "TO", "TON"),
     "tunisia": ("Tunisia", "TN", "TUN"),
     "turkey": ("Turkey", "TR", "TUR"),
+    "turks_and_caicos_islands": ("Turks and Caicos Islands", "TC", "TCA"),
```
Yes, it should be class name.

ShalomDee commented:
I have implemented the following changes:

1. Fixed registry entry class name from "Turks and Caicos Islands" to "TurksAndCaicosIslands"
2. Updated language settings to use en_TC as default with en_US as secondary
3. Added historical context to King's Birthday and National Heritage Day holiday comments
4. Added periods to holiday comments for consistency
5. Removed whitespace from empty line in class docstring
6. Added newlines at end of files in tests/countries and holidays/countries
7. Updated PO file headers and metadata with correct version and author information

PPsyrius commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,76 @@
+
+    def test_country_aliases(self):
+        self.assertAliases(TurksAndCaicosIslands, TC, TCA)
+
```
Please adds:
- test_no_checks

Movable holiday ones (see [`test_indonesia.py`](https://github.com/vacanza/holidays/blob/ce30a785023835e2f2e37d65944cc5aef0b64eec/tests/countries/test_indonesia.py#L243) for example)
- test_commonwealth_day
- test_good_friday
- test_easter_monday
- test_national_youth_day
- test_national_heritage_day
- test_national_day_of_thanksgiving

Fixed holidays with changed requirements that needs testing:
- test_jags_mccartney_day (name change; see [`test_macau.py`](https://github.com/vacanza/holidays/blob/ce30a785023835e2f2e37d65944cc5aef0b64eec/tests/countries/test_macau.py#L477) for reference)
- test_sovereign_birthday (for both king's and queen's birthday)

PPsyrius commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,81 @@
+        # JAGS McCartney Day.
+        self._add_holiday_last_mon_of_may(tr("JAGS McCartney Day"))
+
+        # King's Birthday. Previously Queen's Birthday until 2023.
+        self._add_holiday_2nd_mon_of_jun(tr("King's Birthday"))
```
Same as "JAGS McCartney Day" above

PPsyrius commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,81 @@
+        # National Youth Day.
+        self._add_holiday_last_fri_of_sep(tr("National Youth Day"))
+
+        # National Heritage Day. Previously known as Columbus Day.
+        self._add_holiday_2nd_mon_of_oct(tr("National Heritage Day"))
```
Same as "JAGS McCartney Day" above

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,81 @@
+#           ryanss <ryanssdev@icloud.com> (c) 2014-2017
+#  Website: https://github.com/vacanza/holidays
+#  License: MIT (see LICENSE file)
+
+from gettext import gettext as tr
```
Perhaps worth to keep l10n support? It will be easier to add new languages in future

PPsyrius replied:
Hmm, fine by me

PPsyrius replied:
I've updated all the suggestions according to the scenario where we kept the l10n 🙏

arkid15r commented:
@ShalomDee do you plan to finish your work or we can close this PR?

ShalomDee commented:
> @ShalomDee do you plan to finish your work or we can close this PR?

Yes, I do. I've just updated the tests. It should be okay now.

KJhellico commented:
@ShalomDee, , first of all, you need to:
- merge latest dev branch and resolve conflicts
- update files headers
- refresh .po files (`make l10n`)
- make sure that pre-commit checks are passed (`make pre-commit`)
- make sure that tests are passed (`make test`)

PPsyrius commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,93 @@
+        # Easter Monday.
+        self._add_easter_monday(tr("Easter Monday"))
+
+        # JAGS McCartney Day. Previously National Heroes Day until 2020.
+        if year >= 2020:
+            self._add_holiday_last_mon_of_may(tr("JAGS McCartney Day"))
+        else:
+            self._add_holiday_last_mon_of_may(tr("National Heroes Day"))
```
Do we wanna use its full name ("James Alexander George Smith (JAGS) McCartney Day") instead? [[1]](https://web.archive.org/web/20250429025117/https://www.gov.tc/agc/component/edocman/05-of-2020-public-holidays-amendment-ordinance/viewdocument/1419?Itemid=) [[2]](https://web.archive.org/web/20250429025602/https://www.gov.tc/agc/component/edocman/21-02-public-holidays-ordinance-2/viewdocument/2027?Itemid=)

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,93 @@
+        else:
+            self._add_holiday_2nd_mon_of_oct(tr("Columbus Day"))
+
+        # National Day of Thanksgiving.
+        self._add_holiday_4th_fri_of_nov(tr("National Day of Thanksgiving"))
```
I saw this change, but government annual holiday schedules still had the name "National Day of Thanksgiving", so I was unsure whether to include it.

PPsyrius replied:
Let's keep them as-is for now in that case since Turk and Caicos government didn't use them either 👍

PPsyrius commented on holidays/locale/en_TC/LC_MESSAGES/TC.po:
```diff
@@ -0,0 +1,83 @@
+msgid "Easter Monday"
+msgstr ""
+
+#. JAGS McCartney Day. Previously National Heroes Day until 2020.
+msgid "JAGS McCartney Day"
+msgstr ""
+
+#. JAGS McCartney Day. Previously National Heroes Day until 2020.
```
Running `make l10n` again should take care of the code comment updates for you automatically 👍

PPsyrius reviewed (commented):
Aside from running `make l10n` to automatically update the l10n files, it might also be a good idea to run `make snapshot` as well.

Other than these minor adjustments, this PR looks pretty much ready for merging 👍

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,106 @@
+
+    References:
+        * <https://en.wikipedia.org/wiki/Public_holidays_in_the_Turks_and_Caicos_Islands>
+        * <https://web.archive.org/web/20250429025117/https://www.gov.tc/agc/component/edocman/05-of-2020-public-holidays-amendment-ordinance/viewdocument/1419?Itemid=>
+        * <https://web.archive.org/web/20250429025602/https://www.gov.tc/agc/component/edocman/21-02-public-holidays-ordinance-2/viewdocument/2027?Itemid=>
```
```suggestion
        * [Ordinance 5 of 2020](https://web.archive.org/web/20250429025117/https://www.gov.tc/agc/component/edocman/05-of-2020-public-holidays-amendment-ordinance/viewdocument/1419?Itemid=)
        * [Public Holidays Ordinance](https://web.archive.org/web/20250429025602/https://www.gov.tc/agc/component/edocman/21-02-public-holidays-ordinance-2/viewdocument/2027?Itemid=)
```

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+class TestTC(CommonCountryTests, TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass(TurksAndCaicosIslands, years=range(2010, 2025))
```
```suggestion
        super().setUpClass(TurksAndCaicosIslands, years=range(1963, 2050))
```

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+
+    def test_country_aliases(self):
+        self.assertAliases(TurksAndCaicosIslands, TC, TCA)
+
```
```suggestion

    def test_no_holidays(self):
        self.assertNoHolidays(TurksAndCaicosIslands(years=1962))

```

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+            "2023-03-13",
+            "2024-03-11",
+            "2025-03-10",
+        )
```
```suggestion
        )
        self.assertHolidayName(name, range(1963, 2050))
```
And the same for all simlar holidays.

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+        name_pre_2020 = "National Heroes Day"
+        name_2020 = "JAGS McCartney Day"
+
+        self.assertHolidayName(
```
```suggestion
        self.assertHolidayName(
            name_pre_2020,
            "2015-05-25",
            "2016-05-30",
            "2017-05-29",
            "2018-05-28",
            "2019-05-27",
        )
        self.assertHolidayName(
```

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+            "2025-05-26",
+        )
+
+        self.assertHolidayName(name_pre_2020, range(2010, 2020))
+        self.assertHolidayName(name_2020, range(2020, 2026))
+        self.assertNoHolidayName(name_pre_2020, range(2020, 2026))
+        self.assertNoHolidayName(name_2020, range(2010, 2020))
```
```suggestion
        self.assertHolidayName(name_pre_2020, range(1963, 2020))
        self.assertHolidayName(name_2020, range(2020, 2050))
        self.assertNoHolidayName(name_pre_2020, range(2020, 2050))
        self.assertNoHolidayName(name_2020, range(1963, 2020))
```
And the same for rest of holidays with similar name changes.

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+        )
+
+    def test_2023(self):
+        self.assertHolidayDates(
+            TurksAndCaicosIslands(years=2023),
+            ("2023-01-01"),
+            ("2023-03-13"),
+            ("2023-04-07"),
+            ("2023-04-10"),
+            ("2023-05-29"),
+            ("2023-06-12"),
+            ("2023-08-01"),
+            ("2023-09-29"),
+            ("2023-10-09"),
+            ("2023-11-24"),
+            ("2023-12-25"),
+            ("2023-12-26"),
```
```suggestion
        self.assertHolidays(
            TurksAndCaicosIslands(years=2023),
            ("2023-01-01", "New Year's Day"),
            ("2023-03-13", "Commonwealth Day"),
            ("2023-04-07", "Good Friday"),
            ("2023-04-10", "Easter Monday"),
            ("2023-05-29", "JAGS McCartney Day"),
            ("2023-06-12", "King's Birthday"),
            ("2023-08-01", "Emancipation Day"),
            ("2023-09-29", "National Youth Day"),
            ("2023-10-09", "National Heritage Day"),
            ("2023-11-24", "National Day of Thanksgiving"),
            ("2023-12-25", "Christmas Day"),
            ("2023-12-26", "Boxing Day"),
```

KJhellico replied:
Still need tests for the rest of the holidays (New Year's Day, etc.)

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,219 @@
+            ("2023-12-26"),
+        )
+
+    def test_2025(self):
+        self.assertHolidayDates(
+            TurksAndCaicosIslands(years=2025),
+            ("2025-01-01"),
+            ("2025-03-10"),
+            ("2025-04-18"),
+            ("2025-04-21"),
+            ("2025-05-26"),
+            ("2025-06-09"),
+            ("2025-08-01"),
+            ("2025-09-26"),
+            ("2025-10-13"),
+            ("2025-11-28"),
+            ("2025-12-25"),
+            ("2025-12-26"),
+        )
+
```
```suggestion
```

ShalomDee commented:
Hey, quick question on the King's Birthday in Turks and Caicos:

There's ambiguity starting in 2023+, where the holiday date seems to be determined by government proclamation rather than a fixed rule (e.g., 2025 uses June 23 instead of the 2nd Monday). Do you want to:

- Keep it predictable/testable — stick to 2nd Monday in June (like the current implementation), or
- Go for real-world accuracy — use a manual override or lookup per year?

If option 2, how should that be handled in the main file and tests? Should we fall back to 2nd Monday and override specific years?

KJhellico commented:
> If option 2, how should that be handled in the main file and tests? Should we fall back to 2nd Monday and override specific years?

Yes, option 2, and yes, in the main file and tests.

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,294 @@
+            ("2023-12-26", "Boxing Day"),
+        )
+
+    def test_2025(self):
+        self.assertHolidays(
+            TurksAndCaicosIslands(years=2025),
+            ("2025-01-01", "New Year's Day"),
+            ("2025-03-10", "Commonwealth Day"),
+            ("2025-04-18", "Good Friday"),
+            ("2025-04-21", "Easter Monday"),
+            ("2025-05-26", "JAGS McCartney Day"),
+            ("2025-06-09", "King's Birthday"),
+            ("2025-08-01", "Emancipation Day"),
+            ("2025-09-26", "National Youth Day"),
+            ("2025-10-13", "National Heritage Day"),
+            ("2025-11-28", "National Day of Thanksgiving"),
+            ("2025-12-25", "Christmas Day"),
+            ("2025-12-26", "Boxing Day"),
+        )
+
```
```suggestion
```
As 2025 is already has in `test_l10n_default`.

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,106 @@
+            else tr("National Heroes Day")
+        )
+
+        # Sovereign's Birthday.
+        name = tr("King's Birthday") if self._year >= 2023 else tr("Queen's Birthday")
+        if self._year >= 2023 and self._year in {2025}:
+            self._add_holiday(name, (JUN, 23))
```
```suggestion
        name = (
            # King's Birthday.
            tr("King's Birthday")
            if self._year >= 2023
            # Queen's Birthday.
            else tr("Queen's Birthday")
        )
        dates_obs = {
            2022: (JUN, 3),
            2023: (JUN, 19),
            2024: (JUN, 17),
            2025: (JUN, 23),
        }
        if self._year in dates_obs:
            self._add_holiday(name, dates_obs.get(self._year))
```

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,106 @@
+
+    References:
+        * [Wikipedia](https://en.wikipedia.org/wiki/Public_holidays_in_the_Turks_and_Caicos_Islands)
+        * [Ordinance 5 of 2020](https://web.archive.org/web/20250429025117/https://www.gov.tc/agc/component/edocman/05-of-2020-public-holidays-amendment-ordinance/viewdocument/1419?Itemid=)
+        * [Public Holidays Ordinance](https://web.archive.org/web/20250429025602/https://www.gov.tc/agc/component/edocman/21-02-public-holidays-ordinance-2/viewdocument/2027?Itemid=)
```
```suggestion
        * [Public Holidays Ordinance, rev. 2014](https://web.archive.org/web/20250210082429/https://gov.tc/agc/component/edocman/21-02-public-holidays-ordinance-2/viewdocument/599?Itemid=)
        * [Ordinance 5 of 2020](https://web.archive.org/web/20250429025117/https://www.gov.tc/agc/component/edocman/05-of-2020-public-holidays-amendment-ordinance/viewdocument/1419?Itemid=)
        * [Public Holidays Ordinance, rev. 2021](https://web.archive.org/web/20250429025602/https://www.gov.tc/agc/component/edocman/21-02-public-holidays-ordinance-2/viewdocument/2027?Itemid=)
        * [2017](https://www.facebook.com/photo/?fbid=1137860329642985&set=a.349345645161128)
```

KJhellico replied:
According to Public Holidays Act (at least from 1980 revision),
> The days mentioned in the Schedule to this Ordinance shall be public holidays provided that if any of those holidays falls on a Saturday or Sunday the following Monday shall be a public holiday.
This need to be implemented as common "observed holidays" feature.

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,125 @@
+        # Christmas Day.
+        self._add_observed(self._add_christmas_day(tr("Christmas Day")))
+
+        # Boxing Day.
+        self._add_observed(self._add_christmas_day_two(tr("Boxing Day")))
```
@coderabbitai, there is no `_add_boxing_day` method.

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,126 @@
+from holidays.groups import ChristianHolidays, InternationalHolidays
+from holidays.observed_holiday_base import (
+    ObservedHolidayBase,
+    SAT_SUN_TO_NEXT_MON,
```
```suggestion
    SAT_SUN_TO_NEXT_MON,
    SAT_SUN_TO_NEXT_MON_TUE,
```

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,126 @@
+
+    country = "TC"
+    default_language = "en_TC"
+    # Separated from Jamaica in 1962.
+    start_year = 1963
+    supported_languages = ("en_TC", "en_US")
+    observed_label = "%s (observed)"
```
```suggestion
    # %s (observed).
    observed_label = "%s (observed)"
    # Separated from Jamaica in 1962.
    start_year = 1963
    supported_languages = ("en_TC", "en_US")
```
`observed_label` should have l10n comment and be placed in alphabetical order.

KJhellico commented on holidays/countries/turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,126 @@
+        # National Day of Thanksgiving.
+        self._add_holiday_4th_fri_of_nov(tr("National Day of Thanksgiving"))
+
+        # Christmas Day.
+        self._add_observed(self._add_christmas_day(tr("Christmas Day")))
+
+        # Boxing Day.
+        self._add_observed(self._add_christmas_day_two(tr("Boxing Day")))
```
```suggestion
        self._add_observed(
            # Christmas Day.
            self._add_christmas_day(tr("Christmas Day")),
            rule=SAT_SUN_TO_NEXT_MON_TUE,
        )

        self._add_observed(
            # Boxing Day.
            self._add_christmas_day_two(tr("Boxing Day")),
            rule=SAT_SUN_TO_NEXT_MON_TUE,
        )
```
As we can see in 2021, when December 26-26 falls on a Saturday-Sunday, non-working days are 27-28.

KJhellico commented on tests/countries/test_turks_and_caicos_islands.py:
```diff
@@ -0,0 +1,240 @@
+    def test_no_holidays(self):
+        self.assertNoHolidays(TurksAndCaicosIslands(years=1962))
+
+    def test_new_years_day(self):
+        self.assertHolidayName("New Year's Day", (f"{year}-01-01" for year in range(1963, 2050)))
+
```
For holidays that have "observed" feature, we need to test observed dates (look at Trinidad and Tobago, for example).

arkid15r commented:
@ShalomDee  thank you for your hard work on this! Do you think it'd be feasible to complete the work before our next version release May 19?

ShalomDee commented:
> @ShalomDee thank you for your hard work on this! Do you think it'd be feasible to complete the work before our next version release May 19?

Hi @arkid15r, thank you for checking in. I wasn’t able to wrap this up before the release since I’ve just started a new internship and was finishing onboarding last week. That said, I’ll aim to circle back to this shortly and get it over the line.

KJhellico commented on holidays/locale/en_US/LC_MESSAGES/TC.po:
```diff
@@ -0,0 +1,88 @@
+#
+msgid ""
+msgstr ""
+"Project-Id-Version: Holidays 0.73\n"
```
```suggestion
"Project-Id-Version: Holidays 0.75\n"
```

KJhellico reviewed (approved):
LGTM! 👍

arkid15r reviewed (commented):
Congrats @ShalomDee and thanks for finishing this! 👍
